### ISSUE TITLE
Battle upgrade math

### ISSUE THREAD

**Issue by Revenate39** [OpenThrone/OpenThrone#212]
I just did the math and think I'm being short changed on my offensive stat gain.

Some data to begin with to make sure I'm not missing anything obvious.
I have 9,720 knights, 7,759 Beserkers = 17,479 total offensive troops
My armory is 20K across all tier 2 items
I have 15K Steeds
I have 24 PP's in offence

My original Offence stat was 8,134,646, I purchased 82 steeds for the correct amount of gold. Each steed provides 200 offence. 200 x 82 = 16,400 at an absolute minimum (no bonuses, PP's, etc) I should see an increase to 8,151,046. However no change, Offence stays at the original value.

Converting 1 beserker to knight sees a decrease of 37 Offence, so I don't think anything weird is going on there. 

I sold 5,280 steeds, to have just enough to have 1 steed for each of my knights, and the offence value changed to 8,134,361, a decrease of 248. 

I sold 7,241 steeds, to have just enough to have 1 steed for each of my beserkers, and the offence value changed to 7,647,041, a decrease of 487,565

I'm hoping someone may be able to flesh out what is going on behind the scenes, as I can't quite put my finger on where the issue is. Also thanks for leaving the sell value of battle upgrades at buy value, otherwise I would be broke :)

**Comment by uaktags** [comment]
I think you're right, just spent the better part of the last hour walking through code and the calculations. Found 1 issue where we're not using Structure Upgrades at all in the calculation, but also can't explain a discrepancy in my math and the code's output.

Ready for a journey into this nightmare?

This is how we currently calculate (let's agree to just focus on OFFENSE for now)
https://github.com/OpenThrone/OpenThrone/blob/827b8ae561f89e9285bbd5711bbea25bc5f8e4d4/src/models/Users.tsx#L508-L597

So using the numbers provided
```
Units: 9720 level 2 OFF and 7759 Level 3 OFF (17479 total)
Items: 27000 Level 2 items all tiers (thank you for making that simple). We only count 17479 Items though, as items can't help if not equipped by a unit.
```
So with just these numbers raw, we should be at an offense of:
```
Units: 20*9720 + 50*7759 = 158400 + 387950 = 546350 OFF
Items:  50 * 17479 (W) + 12 * 17479 (H) + 38 * 17479 (A) + 12 * 17479 (BO) + 5*17479 (BR) + 25 * 17479 (S) = 2482018 OFF
Total: 3028368 OFF
```
Then looking at Battle Upgrades
```
Steeds: 200 * 15541 = 3108200
Total with Units+Items: 3108200 + 3028368 = 6136568
```
So up to this point, we have just calculated the RAW values of quantity * the default bonus everything provides. We haven't deducted your personal bonuses from Structure Upgrades, PP, or Race/Class (if any)

Player
```
Race: Elf - No bonus for off
Class: Assassin - No bonus for off
PP: 24
Structure Upgrades: level7 = 30%
```
I'd expect then that we should be at 6136568 * 1.54 = 9450315
So right off the bat, we're never taking Structure Upgrades into consideration is 1 issue i see.
Removing SU from the equation: 6136568 * 1.24 = 7609344

For a sanity check:
Right now at this current moment, you had the following:
https://gist.github.com/uaktags/64ff750f216bf6754eb89d5ab41dc950
Units: 10365 l2, 7759 l3 = 18124 units total = 10365*20 + 7759*50 = 595250
Items: 18124 l2 items = 2573608
BUs: 15539 * 200 = 3107800
PP: 24
Total Offense: (2573608+595250+3107800)*1.24 = 7783055
Game calculated: 8424136

So yea, there's definitely a discrepancy to be resolved.

**Comment by uaktags** [comment]
If someone else could just spot check me, i don't think i missed anything else, but i'd like to use that gist example of Rev's stats as the basis for a unit test. I think it's a perfect example to use as it covers all aspects that we'd expect: PP, BattleUpgrades, Units, Items, structures, etc

**Comment by uaktags** [comment]
Okay, i think i have a fix for the structure upgrades, and have forced the battleupgrades usage to be from highest level to lowest.

With the gist's data points, its coming up with a OFFENSE of 9666054
Units: 10365 l2, 7759 l3 = 18124 units total = 1036520 + 775950 = 595250
Items: 18124 l2 items = 2573608
BUs: 15539 * 200 = 3107800
StructureUpgrade = 30%
PP = 24%
Total Offense: (2573608+595250+3107800)*1.54 = 9666054

So math checks out now.

Before conversion 9666054
after conversion 9666100
difference in offense 46
So from that starting point, converting 1 Knight to Bererker sees a 46 OFFENSE increase (30 OFFENSE diff for Berserker + 54%)
so that checks out. You saw 37 because 30 * 24%.

So the issue was all around the BattleUpgrades like you noticed and the lack of Structure Upgrades.

**Comment by uaktags** [comment]
@Revenate39 the fix for these calculations are in place now. 
Units only use a single BattleUpgrade, and they take the highest level first. 
So Catapults and War Elephants will take priority over level 1 battleupgrades. 
Offense/Defense is updated after every transaction and every 30minutes.

**Comment by JOHNNYGAZER** [comment]
@uaktags Apologies for the "unintelligent" title on the previous issue submitted. I meant to express that my army was being unintelligent by using their guard towers after I rolled in some shiny new catapults for them. It was late when I typed that and I didn't communicate it well.

I went to test this out today and noticed that the defense stat didn't update after the battle upgrade transaction, but instead updated after making a transaction in the armory. The "update after transaction" function could probably be copied over and reused in the battle upgrades page I would imagine.

Also, I can confirm that as expected my archers switched to the catapults without needing to sell any guard towers.
